### PR TITLE
Postgres: improve $__timeFilter macro

### PR DESCRIPTION
**Pull request by svenklemm**
This changes the `__timeFilter` macro expansion from `extract(epoch from column) between unixfrom and unixto` to `column between from and to`. The conversion of from and to to datetime happens in the backend instead of the database. 
The `$__timeFrom` and `$__timeTo` macros got changed accordingly.

This version of the timeFilter macro expansion should no longer prevent index-usage on the time column since no function is called on column itself.

#11578
#10096

### CODE REVIEW

**Comment by marefr** [comment]
I have been thinking about the __timeFilter for some time. Wouldn't the optimal solution for all three sql data sources (postgres, mysql, mssql) be to skip the conversion to epoch in generated sql and instead format the time range in go to a compatible datetime/timestamp string to be used in generated sql. Generated sql could then be similar to:

`date_time_column BETWEEN '2018-04-12 00:00' AND '2018-04-12 23:59'`

If you have an index on your datetime/timestamp this would in my head be the best from a performance perspective. Thoughts @svenklemm?

**Comment by svenklemm** [comment]
This is kinda what the patch does except that the unix timestamp -> database time conversion is happening in the database. The reason i chose this approach is cause i trust the database more to do the right thing regarding timezones. The performance should be similar for both cases. The big performance drawback of the current version is that the expression has to be evaluated for every row before rows can be excluded from the resultset (unless you have a functional index on the time column).

I can change the patch to do the conversion on the go side it would certainly make the patch simpler and the datasource wouldnt need to have a version.

**Comment by marefr** [comment]
@svenklemm nice! If you could rebase this against master I can take it for a spin/review it and hopefully merge it. Would like to fix this for mysql and mssql as well, but we can open specific issues for those and fix later.

File: pkg/tsdb/postgres/macros.go [hunk 1]
```diff
@@ -79,11 +79,11 @@ func (m *PostgresMacroEngine) evaluateMacro(name string, args []string) (string,
 		}
 		return fmt.Sprintf("extract(epoch from %s) as \"time\"", args[0]), nil
 	case "__timeFilter":
-		// dont use to_timestamp in this macro for redshift compatibility #9566
 		if len(args) == 0 {
 			return "", fmt.Errorf("missing time column argument for macro %v", name)
 		}
-		return fmt.Sprintf("extract(epoch from %s) BETWEEN %d AND %d", args[0], m.TimeRange.GetFromAsSecondsEpoch(), m.TimeRange.GetToAsSecondsEpoch()), nil
+
+		return fmt.Sprintf("%s BETWEEN '%s' AND '%s'", args[0], m.TimeRange.MustGetFrom().UTC().Format(time.RFC3339), m.TimeRange.MustGetTo().UTC().Format(time.RFC3339)), nil
```
Comment (marefr) [inline pkg/tsdb/postgres/macros.go]: A bit unsure of the `time.RFC3339`. Would that format be supported by all postgres versions? Or would `'<datetime formated>'::date` be better?

Also, would like `m.TimeRange.MustGetFrom().UTC()` to be encapsulated in `m.TimeRange.GetFromAsTimeUTC()` or something similar.

File: public/app/plugins/datasource/postgres/module.ts [hunk 2]
```diff
@@ -8,8 +8,9 @@ class PostgresConfigCtrl {
 
   /** @ngInject **/
   constructor($scope) {
-    this.current.jsonData.sslmode = this.current.jsonData.sslmode || 'require';
+    this.current.jsonData.sslmode = this.current.jsonData.sslmode || 'verify-full';
```
Comment (marefr) [inline public/app/plugins/datasource/postgres/module.ts]: A bit unsure why this change is needed or if it was mistakenly added?

**Review by marefr — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/postgres/module.ts (hunk 2 above)
Comment (svenklemm) [reply public/app/plugins/datasource/postgres/module.ts]: This is unrelated i'll do a separate pr for this.

File: pkg/tsdb/postgres/macros.go (hunk 1 above)
Comment (svenklemm) [reply pkg/tsdb/postgres/macros.go]: I checked on 8.0.2 which is 13 years old and what redshift is based on and it worked there. So its safe to assume any recent postgres supports it. I choose this format cause its a standard and has the timezone in it aswell.

File: pkg/tsdb/postgres/macros.go [hunk 3]
```diff
@@ -79,11 +79,11 @@ func (m *PostgresMacroEngine) evaluateMacro(name string, args []string) (string,
...
 		return fmt.Sprintf("extract(epoch from %s) as \"time\"", args[0]), nil
 	case "__timeFilter":
-		// don't use to_timestamp in this macro for redshift compatibility #9566
 		if len(args) == 0 {
 			return "", fmt.Errorf("missing time column argument for macro %v", name)
 		}
-		return fmt.Sprintf("extract(epoch from %s) BETWEEN %d AND %d", args[0], m.TimeRange.GetFromAsSecondsEpoch(), m.TimeRange.GetToAsSecondsEpoch()), nil
+
+		return fmt.Sprintf("%s BETWEEN '%s' AND '%s'", args[0], m.TimeRange.GetFromAsTimeUTC().Format(time.RFC3339), m.TimeRange.GetToAsTimeUTC().Format(time.RFC3339)), nil
 	case "__timeFrom":
```
Comment (marefr) [inline pkg/tsdb/postgres/macros.go]: For consistency I think that `$__timeFrom` and `$__timeTo` should return a value formatted as`'2006-01-02T15:04:05Z07:00'` like `$__timeFilter` instead of `to_timestamp(<epoch>)`. What do you think?

**Review by marefr — CHANGES_REQUESTED** [review]
This looks great! Have a minor comment that I would like your feedback on.

Also, the PR description could be changed now since scope has changed a bit. Is this PR still fixing #10096?

File: pkg/tsdb/postgres/macros.go (hunk 3 above)
Comment (svenklemm) [reply pkg/tsdb/postgres/macros.go]: I agree `$__timeFrom` and `$__timeTo` should be adjusted. If someone really needs the old functionality you can still get it with the unix timestamp macros and pass those into to_timestamp but they should evaluate to the same thing anyway.

**Comment by svenklemm** [comment]
@marefr Ok i adjusted the description of the pr. Regarding  #10096 i still think it fixes the underlying issue. The suggested solution to make the macro configurable was not implemented but the suboptimal queries got fixed.

**Review by marefr — APPROVED** [review]
:+1:

**Comment by marefr** [comment]
Thank you for contributing to Grafana!
